File: src/ide/webstorm/src/main/java/com/point/i18n/webstorm/services/TranslateService.java
package com.point.i18n.webstorm.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;

/**
 * Service for translation using DeepL API.
 */
public class TranslateService {
    private static final String API_URL = "https://api-free.deepl.com/v2/translate";
    private static final List<String> BETA_LANGUAGES = Arrays.asList("UZ");
    
    private final ConfigService configService;
    private final ObjectMapper objectMapper;
    
    public TranslateService(ConfigService configService) {
        this.configService = configService;
        this.objectMapper = new ObjectMapper();
        this.objectMapper.configure(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }
    
    public String getApiKey() throws IllegalStateException {
        String apiKey = configService.getDeepLApiKey();
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new IllegalStateException("DeepL API ключ не настроен");
        }
        return apiKey;
    }
    
    public String translate(String text, String targetLanguage) throws IOException {
        if (text == null || text.trim().isEmpty()) {
            return "";
        }
        
        String apiKey = getApiKey();
        String sourceLanguage = "RU"; // Всегда переводим с русского
        String targetLang = targetLanguage.toUpperCase();
        
        boolean isBetaLanguage = BETA_LANGUAGES.contains(targetLang);
        
        // Формируем form-data
        StringBuilder formData = new StringBuilder();
        formData.append("text=").append(URLEncoder.encode(text, StandardCharsets.UTF_8));
        formData.append("&source_lang=").append(sourceLanguage);
        formData.append("&target_lang=").append(targetLang);
        if (isBetaLanguage) {
            formData.append("&enable_beta_languages=1");
        }
        
        URL url = new URL(API_URL);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Authorization", "DeepL-Auth-Key " + apiKey);
        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
        connection.setDoOutput(true);
        
        try (java.io.OutputStream os = connection.getOutputStream()) {
            byte[] input = formData.toString().getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }
        
        int responseCode = connection.getResponseCode();
        String responseText;
        
        // Читаем ответ
        try (java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(
                    responseCode >= 200 && responseCode < 300 
                        ? connection.getInputStream() 
                        : connection.getErrorStream(), 
                    StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            responseText = response.toString();
        }
        
        if (responseCode != HttpURLConnection.HTTP_OK) {
            String errorMessage = "Ошибка перевода: " + responseCode;
            try {
                TranslateErrorResponse errorResponse = objectMapper.readValue(responseText, TranslateErrorResponse.class);
                if (errorResponse.message != null && !errorResponse.message.isEmpty()) {
                    errorMessage = errorResponse.message;
                }
            } catch (Exception e) {
                // Игнорируем ошибку парсинга
            }
            throw new IOException(errorMessage);
        }
        
        TranslateResponse translateResponse = objectMapper.readValue(responseText, TranslateResponse.class);
        
        if (translateResponse.translations == null || translateResponse.translations.isEmpty()) {
            throw new IOException("Пустой ответ от DeepL API");
        }
        
        return translateResponse.translations.get(0).text;
    }
    
    public TranslationResult translateToEnAndUz(String ruText) throws IOException {
        if (ruText == null || ruText.trim().isEmpty()) {
            return new TranslationResult("", "");
        }
        
        String en = translate(ruText, "EN");
        
        String uz = "";
        try {
            uz = translate(ruText, "UZ");
        } catch (Exception e) {
            // Если перевод на узбекский не удался, оставляем пустую строку
            System.out.println("Ошибка перевода на узбекский язык: " + e.getMessage());
        }
        
        return new TranslationResult(en, uz);
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class TranslateResponse {
        public java.util.List<Translation> translations;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class Translation {
        public String text;
        public String detected_source_language;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class TranslateErrorResponse {
        public String message;
    }
    
    public static class TranslationResult {
        public final String en;
        public final String uz;
        
        public TranslationResult(String en, String uz) {
            this.en = en;
            this.uz = uz;
        }
    }
}

File: src/services/translate.service.ts
import * as vscode from 'vscode';
import fetch from 'node-fetch';
import { URLSearchParams } from 'url';

export class TranslateService {
    private readonly configSection = 'i18nRemote';
    private readonly apiKeyConfigKey = 'deepLApiKey';
    private readonly apiUrl = 'https://api-free.deepl.com/v2/translate';
    
    // Список бета-языков DeepL
    private readonly betaLanguages = ['UZ'];

    async getApiKey(): Promise<string | null> {
        const config = vscode.workspace.getConfiguration(this.configSection);
        return config.get<string>(this.apiKeyConfigKey) || null;
    }

    async ensureApiKey(): Promise<string> {
        let apiKey = await this.getApiKey();
        
        if (!apiKey) {
            const action = await vscode.window.showErrorMessage(
                'DeepL API ключ не настроен. Укажите API ключ для работы с переводом.',
                'Настроить'
            );
            
            if (action === 'Настроить') {
                await vscode.commands.executeCommand('i18nRemote.configDeepLApiKey');
                apiKey = await this.getApiKey();
                
                if (!apiKey) {
                    throw new Error('DeepL API ключ не настроен');
                }
            } else {
                throw new Error('DeepL API ключ не настроен');
            }
        }
        
        return apiKey;
    }

    async translate(text: string, targetLanguage: string): Promise<string> {
        if (!text || !text.trim()) {
            return '';
        }

        const apiKey = await this.ensureApiKey();
        const sourceLanguage = 'RU'; // Всегда переводим с русского

        // DeepL использует коды языков в верхнем регистре
        const targetLang = targetLanguage.toUpperCase();
        
        // Проверяем, является ли целевой язык бета-языком
        const isBetaLanguage = this.betaLanguages.includes(targetLang);

        const formData = new URLSearchParams();
        formData.append('text', text);
        formData.append('source_lang', sourceLanguage);
        formData.append('target_lang', targetLang);
        
        // Добавляем параметр для бета-языков
        if (isBetaLanguage) {
            formData.append('enable_beta_languages', '1');
        }

        const response = await fetch(this.apiUrl, {
            method: 'POST',
            headers: {
                'Authorization': `DeepL-Auth-Key ${apiKey}`,
                'Content-Type': 'application/x-www-form-urlencoded',
            },
            body: formData.toString()
        });

        if (!response.ok) {
            const errorText = await response.text();
            let errorMessage = `Ошибка перевода: ${response.status}`;
            
            try {
                const errorData = JSON.parse(errorText);
                if (errorData.message) {
                    errorMessage = errorData.message;
                }
            } catch {
                // Игнорируем ошибку парсинга
            }
            
            throw new Error(errorMessage);
        }

        const data = await response.json() as {
            translations: Array<{
                text: string;
                detected_source_language?: string;
            }>;
        };

        if (!data.translations || data.translations.length === 0) {
            throw new Error('Пустой ответ от DeepL API');
        }

        return data.translations[0].text;
    }

    async translateToEnAndUz(ruText: string): Promise<{ en: string; uz: string }> {
        if (!ruText || !ruText.trim()) {
            return { en: '', uz: '' };
        }

        try {
            // Переводим на английский
            const en = await this.translate(ruText, 'EN');
            
            // Переводим на узбекский (бета-язык в DeepL)
            let uz = '';
            try {
                uz = await this.translate(ruText, 'UZ');
            } catch (uzError: any) {
                // Если перевод на узбекский не удался, оставляем пустую строку
                console.log('Ошибка перевода на узбекский язык:', uzError.message);
            }
            
            return { en, uz };
        } catch (error: any) {
            const errorMessage = error.message || 'Ошибка перевода';
            throw new Error(errorMessage);
        }
    }
}

